saleor/checkout/complete_checkout.py

```diff
@@ -545,14 +550,93 @@ def _process_payment(
                 channel_slug=channel_slug,
             )
         payment.refresh_from_db()
+        payment.complete_order = complete_order
+        payment.save(update_fields=["complete_order"])
         if not txn.is_success:
             raise PaymentError(txn.error)
     except PaymentError as e:
-        release_voucher_usage(order_data)
+        if order_data:
+            release_voucher_usage(order_data)
+        call_payment_refund_or_void(channel_slug, payment, manager)
         raise ValidationError(str(e), code=CheckoutErrorCode.PAYMENT_ERROR.value)
     return txn
 
 
+def _complete_checkout_payment(
+    channel_slug: str,
+    payment: "Optional[Payment]",
+    manager: "PluginsManager",
+    payment_data,
+    store_source,
+    user,
+    order_data=None,
+    complete_order: bool = False,
+):
+    """Complete processing a payment.
+
+    If no payment was provided, it means that it was already processed
+    by this function as part of checkoutPaymentComplete mutation.
+    """
+    customer_id = None
+    if not payment:
+        return {}, False
+
+    if user.is_authenticated:
+        customer_id = fetch_customer_id(user=user, gateway=payment.gateway)
+
+    txn = _process_payment(
```

Comment:
What does `txn` mean?

Reply:
It's a shortcut for transaction.

Reply:
I don't want to excuse myself, but it's an extracted method (https://github.com/mirumee/saleor/pull/8001/files/2ccb78ffba77d98d8364fbe443340c10bfbb6f40#diff-27b3f5d7b64efa6b9211eb632292fa7d6f6e031c8a5a38581da6391a901b5d0eL598) and it's used throughout Saleor already (e.g. in pytest fixture names) :bowtie: I wouldn't change the term right now even if it could be confusing to some people